framework/src/modules/dpos_v2/auxiliary.ts

```diff
@@ -0,0 +1,64 @@
+import { BlockAfterExecuteContext } from '../../node/state_machine/types';
+import { MODULE_ID_DPOS, STORE_PREFIX_DELEGATE } from './constants';
+import { delegateStoreSchema } from './schemas';
+import { DelegateAccount, ModuleConfig, ValidatorsAPI } from './types';
+
+export const updateDelegateProductivity = async ({
+	validatorsAPI,
+	previousTimestamp,
+	context,
+	moduleConfig,
+}: {
+	validatorsAPI: ValidatorsAPI;
+	previousTimestamp: number;
+	context: BlockAfterExecuteContext;
+	moduleConfig: ModuleConfig;
+}) => {
+	const { header, getAPIContext, getStore } = context;
+
+	const newHeight = header.height;
+	const apiContext = getAPIContext();
+	const missedBlocks = await validatorsAPI.getGeneratorsBetweenTimestamps(
+		apiContext,
+		previousTimestamp,
+		header.timestamp,
+	);
+
+	const generatorAtPreviousTimestamp = await validatorsAPI.getGeneratorAtTimestamp(
+		apiContext,
+		previousTimestamp,
+	);
+	const generatorAtCurrentTimestamp = await validatorsAPI.getGeneratorAtTimestamp(
+		apiContext,
+		header.timestamp,
+	);
+
+	missedBlocks[generatorAtPreviousTimestamp.toString('hex')] -= 1;
```

Comment:
I think we can adjust the API to not to  include the previous timestamp.
ie) the API should return result  from `(from, to]`. @Incede @ricott1  ?